### ISSUE TITLE
Another BUG with Chest (184465) - can't be opened (BB #471)

### ISSUE THREAD

**Issue by OregonAsari** [OregonCore/OregonCore#471]
This issue was migrated from bitbucket.
**Original Reporter:** PadreWoW
**Original Date:** 14.08.2013 12:06:55 GMT+0000
**Original Priority:** major
**Original Type:** bug
**Original State:** resolved
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/471

<hr>

Another BUG with Chest (184465) - can't be opened

Nothing happens when you use Cache of the Legion (or it's key on it)

No loot, and no loot-window

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** PadreWoW
**Original Date:** 14.08.2013 13:38:13 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/471/#comment-5610929

<hr>

+Benedict's Chest

+Captain's Chest

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** PadreWoW
**Original Date:** 14.08.2013 13:38:42 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/471/#comment-5610938

<hr>

I think all chests with keys are bugged

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** PadreWoW
**Original Date:** 14.08.2013 13:49:31 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/471/#comment-5611081

<hr>

+Postboxes (all) in Stratholme

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** PadreWoW
**Original Date:** 15.08.2013 11:59:46 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/471/#comment-5623651

<hr>

The problem is here:

```
TakeCastItem();

    if (m_spellState != SPELL_STATE_CASTING)
        finish(true); // successfully finish spell cast (not last in case autorepeat or channel spell)
}
```

When the spell ends, the chest closes

But i can't fix it
There should be something like this:
Key must become unselectable (gray), until player takes loot, and disappear after player closes loot window

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Yooki
**Original Date:** 29.12.2013 17:55:47 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/471/#comment-7523335

<hr>

The problem was in the 'TakeCastItem' function.

It would seem that when the key is removed from the inventory at that particular point the chest refuses to open. A small note would be that this does not affect the chests that need to be lockpicked in order to open them. I've written up a simple fix, although it might not be the best, one could write but the chests should open properly now and the loot window should appear as intended.

I have checked out the postboxes in Stratholme you mentioned, and unfortunately I do not know how to take care of that problem at this moment, mostly because I am not familiar with the Stratholme instance. If you could provide me more information about the postboxes and what they should drop I might be able to do something.

The commit addressing this issue is available in my OregonCore Fork, if anyone is interested about this.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** PadreWoW
**Original Date:** 29.12.2013 20:09:38 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/471/#comment-7524104

<hr>

Fixed by LorUsagi long time ago

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** PadreWoW
**Original Date:** 29.12.2013 20:10:22 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/471/#comment-7524110

<hr>

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Yooki
**Original Date:** 30.12.2013 07:40:00 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/471/#comment-7527481

<hr>

I think it would be good practice to publish solutions to these issues when they are found.

Since none of these updates are being merged into the official repository, other users/developers that stumble upon these issues are only left to guess as to how exactly are they resolved and some of them might even want to apply these fixes to their cores if they are experiencing the same problem or if they simply want to remove as much bugs as possible from their project.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 30.12.2013 16:28:34 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/471/#comment-7531879

<hr>

Mines probably quite different from yours mate. I don't have a lot of experience in this sort of thing.

```
#!c++

         }

     }



+   // Check for unlocking spells, force skip a check and take the key.

+   for (int i = 0; i < 3; i++)

+   {

+       if(m_spellInfo->Effect[i] == SPELL_EFFECT_OPEN_LOCK_ITEM)

+       {

+           TakeCastItem();

+           skipCheck = true;

+       }

+   }

+

     // triggered cast called from Spell::prepare where it was already checked

     if (!skipCheck)

     {


```

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Yooki
**Original Date:** 31.12.2013 16:59:39 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/471/#comment-7540719

<hr>

This should work in most cases, but in some cases the spell might have a secondary or tertiary effect. There are also attributes that better define where and when the spell should or should not be used. Such effects and attributes could have special requirements or restrictions, and while it might not seem important at the first glance, it could render some custom made spells inefficient and inadequate in their designated function.

I have modified the fix a little bit (with your help) to be a better reflection of what I was talking about. It was inconsistent for it to be in the 'canCast' function, which should have been 'const' to begin with.

```
#!c++
         if (m_spellInfo->Effect[j] == 0)

             continue;



+       // Remove the key from player's inventory here to make sure that

+       // locked chests open and display loot properly.

+       if(m_spellInfo->Effect[j] == SPELL_EFFECT_OPEN_LOCK_ITEM)

+          TakeCastItem();

+

         // apply Send Event effect to ground in case empty target lists

         if (m_spellInfo->Effect[j] == SPELL_EFFECT_SEND_EVENT && !HaveTargetsForEffect(j))

         {
```

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** release_ok
**Original Date:** 02.01.2014 13:21:43 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/471/#comment-7550366

<hr>

Resolved issue #471 - Locked chests can't be opened:

Locked chests were unable to be looted as after an attempt to open them, they
would quickly close and the LootView would never appear.

This however was not the case with locked chests that required lockpicking.
Postboxes in Stratholme still remain broken.

Thanks to LordUsagi for helping me out with this issue.

→ <<cset f006eaad6ad2>>
